lib/curl/logger.ex

```diff
@@ -1,220 +1,51 @@
-  end
-
-  def handle_info(_, state) do
-    {:ok, state}
-  end
-
-  def code_change(_old_vsn, state, _extra) do
-    {:ok, state}
-  end
-
-  def terminate(_reason, _state) do
-    :ok
-  end
-
-  ## Helpers
-
-  defp meet_level?(_lvl, nil), do: true
-
-  defp meet_level?(lvl, min) do
-    Logger.compare_levels(lvl, min) != :lt
-  end
-
-  defp configure(options, state) do
-    config = Keyword.merge(Application.get_env(:logger, __MODULE__), options)
-    Application.put_env(:logger, __MODULE__, config)
-    init(config, state)
-  end
-
-  defp init(config, state) do
-    level = Keyword.get(config, :level)
-    service = Keyword.get(config, :service)
-    device = Keyword.get(config, :device, :user)
-    metadata = Keyword.get(config, :metadata, []) |> configure_metadata()
-    max_buffer = Keyword.get(config, :max_buffer, 32)
-
-    %{
-      state
-      | metadata: metadata,
-        level: level,
-        service: service,
-        device: device,
-        max_buffer: max_buffer
-    }
-  end
-
-  defp configure_metadata(:all), do: :all
-  defp configure_metadata(metadata), do: Enum.reverse(metadata)
-
-  defp log_event(level, msg, ts, md, %{device: device} = state) do
-    output = format_event(level, msg, ts, md, state)
-    %{state | ref: async_io(device, output), output: output}
-  end
-
-  defp buffer_event(level, msg, ts, md, state) do
-    %{buffer: buffer, buffer_size: buffer_size} = state
-    buffer = [buffer | format_event(level, msg, ts, md, state)]
-    %{state | buffer: buffer, buffer_size: buffer_size + 1}
-  end
-
-  defp async_io(name, output) when is_atom(name) do
-    case Process.whereis(name) do
-      device when is_pid(device) ->
-        async_io(device, output)
-
-      nil ->
-        raise "no device registered with the name #{inspect(name)}"
+  def format(level, message, timestamp, metadata) do
+    try do
+      message
+      |> to_map
+      |> add_timestamp(timestamp)
+      |> add_metadata(metadata)
+      |> add_level(level)
+      |> add_service
+      |> to_json
+    rescue
+      _ ->
+        case Poison.encode(%{message: "could not format: #{inspect message}}", level: level, metadata: inspect(metadata)}) do
+          {:ok, message} -> message
+          _ -> ""
+        end
     end
-  end
 
-  defp async_io(device, output) when is_pid(device) do
-    ref = Process.monitor(device)
-    send(device, {:io_request, self(), ref, {:put_chars, :unicode, output}})
-    ref
   end
 
-  defp await_io(%{ref: nil} = state), do: state
-
-  defp await_io(%{ref: ref} = state) do
-    receive do
-      {:io_reply, ^ref, :ok} ->
-        handle_io_reply(:ok, state)
-
-      {:io_reply, ^ref, error} ->
-        handle_io_reply(error, state)
-        |> await_io()
-
-      {:DOWN, ^ref, _, pid, reason} ->
-        raise "device #{inspect(pid)} exited: " <> Exception.format_exit(reason)
+  defp to_map(message) when is_binary(message) do
+    case Poison.decode(message) do
+      {:ok, %{} = json} -> json
```

Comment:
Just to make sure we don't have keys collision and to easily differentiate the message from the rest, I'd still do `{:ok, %{} = json} -> %{message: json}` in this case as well.

Reply:
They need to be merged at the top level for searching in kibana. You could prefix all metadata with `_` or similar